src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -38,18 +59,9 @@ private CSharpSyntaxNode ParseTypeOrPatternForIsOperator()
                     break;
             }
 
-            if (tk == SyntaxKind.IdentifierToken && this.CurrentToken.Text == "_")
-            {
-                // In a pattern, we reserve `_` as a wildcard. It cannot be used (with that spelling) as the
-                // type of a declaration or recursive pattern, nor as a type in an in-type expression starting
-                // in C# 7. The binder will give a diagnostic if
-                // there is a usable symbol in scope by that name. You can always escape it, using `@_`.
-                // TODO(patterns2): Should we use the "contextual keyword" infrastructure for this?
-                var node = _syntaxFactory.DiscardPattern(this.EatToken(SyntaxKind.IdentifierToken));
-                return this.CheckFeatureAvailability(node, MessageID.IDS_FeatureRecursivePatterns);
-            }
-
-            // If it starts with 'nameof(', skip the 'if' and parse as a constant pattern.
+            // We permit a type named `_` on the right-hand-side of an is operator, but not inside of a pattern.
+            bool typeCannotBePattern = tk == SyntaxKind.IdentifierToken && this.CurrentToken.Text == "_";
```

Comment:
>bool typeCannotBePattern = tk == SyntaxKind.IdentifierToken && this.CurrentToken.Text == "_"; [](start = 12, length = 93)

It feels like we could end up parsing ```_``` as a constant pattern rather than a TypeSyntax. Is this intentional?

Reply:
I don't see a way of that occurring. The closest thing that I can see is that `e is _+1` would parse the right-hand-side as a constant pattern. That is intentional.